src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
+                    };
+                })
+                .Where(x => x.ResourceNamePropertySyntax is not null && x.ResourceNamePropertyValue is not null)
+                .GroupBy(x => (x.ResourceTypeFQDN, x.ResourceScope, x.ResourceNamePropertyValue!), ResourceComparer.Instance)
+                .Where(group => group.Count() > 1);
+
+            foreach (var duplicatedResourceGroup in duplicateResources)
+            {
+                var duplicatedResourceNames = string.Join(", ", duplicatedResourceGroup.Select(x => x.ResourceName));
+                foreach (var duplicatedResource in duplicatedResourceGroup)
+                {
+                    yield return DiagnosticBuilder.ForPosition(duplicatedResource.ResourceNamePropertySyntax!).ResourceMultipleDeclarations(duplicatedResourceNames);
+                }
+            }
+
+            var duplicateModules = semanticModel.Root.AllDeclarations
+                .OfType<ModuleSymbol>()
+                .Where(res => res.Type is ModuleType && res.DeclaringModule is ModuleDeclarationSyntax && moduleScopeData.ContainsKey(res))
+                .Select(res =>
+                {
+                    var namePropSyntax = ((ObjectSyntax)res.DeclaringModule.Body).SafeGetPropertyByName(LanguageConstants.ResourceNamePropertyName);
+                    var scopeType = moduleScopeData[res].RequestedScope;
+                    var scopeProperty = ((ObjectSyntax)res.DeclaringModule.Body).SafeGetPropertyByName(LanguageConstants.ResourceScopePropertyName);
+
+                    var r = new
+                    {
+                        ModuleName = res.Name,
+                        ModulePropertyScopeType = scopeType,
+                        ModulePropertyScopeValue = ((scopeProperty?.Value as FunctionCallSyntax)?.Arguments.FirstOrDefault()?.Expression as StringSyntax),
+                        ModulePropertyNameValue = namePropSyntax?.Value as StringSyntax, //TODO: currently limiting check to property values that are strings, although it can be references or other syntaxes
+                        ModulePropertyNameSyntax = namePropSyntax
+                    };
+                    return r;
+                })
+                .Where(x => x.ModulePropertyNameSyntax is not null && x.ModulePropertyNameValue is not null)
+                .GroupBy(x => (x.ModulePropertyScopeType, x.ModulePropertyScopeValue, x.ModulePropertyNameValue!), ModuleComparer.Instance)
+                .Where(group => group.Count() > 1);
+
+            foreach (var duplicatedModuleGroup in duplicateModules)
+            {
+                var duplicatedModuleNames = string.Join(", ", duplicatedModuleGroup.Select(x => x.ModuleName));
+                foreach (var duplicatedModule in duplicatedModuleGroup)
+                {
+                    yield return DiagnosticBuilder.ForPosition(duplicatedModule.ModulePropertyNameSyntax!).ModuleMultipleDeclarations(duplicatedModuleNames);
```

Comment:
Informally we're trying to keep `!` out of the (non-test) code, as it's a potential place for bugs to appear in the future. Unfortunately C# is often not super smart at tracking nullability, but there's usually a way to rewrite so that the compiler figures it out (e.g. rewriting so that the `!` is not necessary). Is that possible here?

Reply:
Will try to avoid it, perhaps with the dedicated class I'll be able to avoid using `!` operator

Reply:
avoided :)